sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/CodeGenInternal/DataFeedDetailPatch.cs

```diff
@@ -14,5 +16,130 @@ internal partial class DataFeedDetailPatch
 
         /// <summary> data feed viewer. </summary>
         public IList<string> Viewers { get; set; }
+
+        // Full qualification must be used so CodeGen skips the generation of this method.
+        // See: https://github.com/Azure/autorest.csharp/issues/793
+        void global::Azure.Core.IUtf8JsonSerializable.Write(Utf8JsonWriter writer)
+        {
+            writer.WriteStartObject();
+
+            if (DataSourceType != default)
+            {
+                writer.WritePropertyName("dataSourceType");
+                writer.WriteStringValue(DataSourceType.ToString());
+            }
```

Comment:
This is the portion we wanted to fix. The service does not accepts a `null` value here anymore. This is NOT a bug on the service side since the swagger does not set `x-nullable` to `true` in this property. We're manually changing the serialization method to add the property only if it has a value.

Reply:
Not sure I understand. Could this be fixed with a swagger transform?

Reply:
The whole problem starts in `dataSourceType` being [a required property in the swagger](https://github.com/Azure/azure-rest-api-specs/blob/bd75cbc7ae9c997f39362ac9d19d557219720bbd/specification/cognitiveservices/data-plane/MetricsAdvisor/preview/v1.0/MetricsAdvisor.json#L4556-L4558):
``` json
"DataFeedDetailPatch": {
  "required": [
    "dataSourceType"
  ],
```

Because of this, CodeGen will send `null` when this property is not set by the user. The service does not accept this property as `null`, though, which is expected, since it's not marked with `"x-nullable": true`. This leads us to a single conclusion: the source type property MUST be set by user.

However, there are scenarios in which it does not make sense to force users to specify a source type. For instance, if they wanted to update only the name of a data feed, they should be able to do:
```cs
var dataFeed = new DataFeed() { Name = "new name" };
client.UpdateDataFeed(dataFeed, "<data-feed-id>");
```
instead of:
```cs
var dataFeed = new DataFeed() { Name = "new name", DataSource = new DataSourceB(params_we_dont_want_to_update)  };
client.UpdateDataFeed(dataFeed, "<data-feed-id>");
```

You may wonder why the service sets it as required, then, if it's not actually required for an update call. [This is part of the swagger specification](https://swagger.io/specification/v2/). `dataSourceType` is a discriminator property:
> The discriminator is the schema property name that is used to differentiate between other schema that inherit this schema. The property name used MUST be defined at this schema and it MUST be in the required property list.

We could, as you suggested, try to use a transform to remove the `dataSourceType` from the list of required properties. This, however, breaks CodeGen, and produces code with compilation errors. It's hard to call this a bug in CodeGen, given that it's a scenario that's not part of the swagger specification.

The simplest solution I could think of is overwriting the serialization method.
https://github.com/Azure/azure-sdk-for-net/blob/5549b7f05dc17f5b75c2bf798ad4b6787043d65d/sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/CodeGenInternal/DataFeedDetailPatch.cs#L26-L30

Originally there was not `if` block here, and a `null` data source type could be sent to the service. Now we're making a check first, and we don't send it if not specified.

Reply:
thanks for the detail explanation. I see. more about swagger specification than anything.